text/0000-kafka-topics-per-tenant.md

```diff
@@ -0,0 +1,121 @@
+
+- Start Date: 2022-08-12
+- RFC PR: (leave this empty)
+- FOLIO Issue: (leave this empty)
+
+# Tenant Collection Kafka Topics
+
+## Summary
+
+The purpose of this RFC is to lay out the reasoning for creating an option to use Kafka topics created for a collection of tenants instead of separate Kafka topics for each tenant. This will reduce the number of Kafka topics and subsequently the number of partitions.
+
+## Motivation
+FOLIO creates Kafka topics and partitions on a per tenant basis for each application concern. There are scaling and cost implications as the number of tenants increases in a FOLIO cluster.
+
+- **Performance**: 
+Similar to a database not having a theoretical limit on the number of rows/tables, Kafka does not have a limit on partition count. A major difference is that a partition is "heavier" than a row/table. Partitions have to be rebalanced, replicated & maintain corresponding open server files for each partition. Partitions have an ongoing administrative cost while online. Creating numerous partitions like they are database rows/tables will spell disaster for an under-provisioned Kafka installation. This is why cloud providers ensure to have limits on partition counts. Provisioning a Kafka installation to support such a high partition count will be wasteful for FOLIO hosting providers. <p>
+On the client side, buffers are created for each partition in memory. So a producer that interacts with very many thousands of partitions will consume more memory. To subscribe to a number of partitions with a regex subscription pattern, Kafka will send the total list of topics to each consumer. A consumer would then filter the list to determine the topics that match the subscription pattern. Sending a total list of topics would require some consensus from all brokers, therefore overhead. Consumers would ask for the full list at regular interval to catch newly added topics.
+- **Cost**: Kafka's use in FOLIO is still limited, more FOLIO modules will continue to employ event driven techniques that will cause an explosion of partitions. For hosting providers that self host Kafka, more brokers or broker capacity will be needed to service load generated by the high number of partitions. For hosting providers who utilize cloud managed Kafka, cloud providers typically charge by partition or set limits on partition count for a particular broker size. For example, Confluent Cloud charges $13/partition/year with a maximum of 4096 partitions for one cluster. AWS charges $3679/year for a broker instance with 4 vCPU and 16GB Memory and a soft limit of 1000 partitions per broker. This does not include extra brokers in other availability zones to support high availability.
+
+## Out of Scope
+- Enabling tenant separation within a module. Tenant separation meaning "How do we prevent a tenant's data from being processed in place of another?", "How can we prevent a module (specific to a tenant) from seeing data from other tenants?"
```

Comment:
I have removed the line.

>Seeing (Kafka) data from other tenants is not only in scope of this RFC, but a key concern of this RFC and should be explained in detail in the RFC.

I am not sure what you are referring to here. Are you implying that seeing Kafka data from other tenants is not possible now and this RFC makes that possible? I would prefer to continue this convo on slack.

Reply:
I explained some reasons why the current state(prior to this RFC) would have to overcome significant hurdles in the related concerns section: `Why Finishing The Temporary Kafka Security Solution Will Have Cascading Effects` to prevent "Seeing (Kafka) data from other tenants".
Others include:
- Changing FOLIO modules to accept multiple kafka credentials for the topics that need to be produced/consumed.
    - Does this make sense if there are no module instances per tenant?
- Needing a single consumer for a topic for each tenant since the authentication identity has to be maintained while the consumer is live. if a FOLIO module needs 10 distinct topics to consume with 10 distinct handlers in the code & 10 tenants exists, there would be 100 topics present in kafka. 100 consumer threads would be needed in each module to . This will cause the number of threads in a module to explode.

I do not want to set a higher standard for this RFC by solving a problem that the current state has not provided a suitable solution. "Seeing (Kafka) data from other tenants" is a drawback that is shared by this RFC and current state(prior to this RFC). Detailing ideas I have about encrypting events is a whole 'nother design that deserves its own document and review iterations. This document should remained focused on the problem at hand: reducing partitions counts.